### ISSUE TITLE
error while building

### ISSUE THREAD

**Issue by idkjs** [janestreet/ocaml_intrinsics#1]
Hey folks,

```sh
sw_vers
ProductName:    macOS
ProductVersion: 11.4
BuildVersion:   20F71
```
This is an m1 mac.

I tried building `core_bench` with `bleeding_edge` installed and got this error while doing so:

```sh
> opam switch create . ocaml-base-compiler.4.11.1 --no-install
> opam repo add janestreet-bleeding https://ocaml.janestreet.com/opam-repository
> eval "$(opam env)"
> opam install core core_kernel textutils
[NOTE] Package core_kernel is already installed (current version is v0.15~preview.124.35+330).
The following actions will be performed:
  ∗ install ocaml_intrinsics v0.15~preview.124.35+330 [required by core]
  ∗ install core             v0.15~preview.124.35+330
  ∗ install textutils        v0.15~preview.124.35+330
===== ∗ 3 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
[ocaml_intrinsics.v0.15~preview.124.35+330] downloaded from https://github.com/janestreet/ocaml_intrinsics/archive/0654b7e6c6cd3818dd07c24ed36cc2399122757c.tar.gz
[textutils.v0.15~preview.124.35+330] downloaded from https://github.com/janestreet/textutils/archive/4dd1d3afb990054e7e42365ac44f660810d24055.tar.gz
[core.v0.15~preview.124.35+330] downloaded from https://github.com/janestreet/core/archive/93dcd36f152b124523dd2584b1fcdcd33d134565.tar.gz

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
[ERROR] The compilation of ocaml_intrinsics failed at "/Users/mando/.opam/opam-init/hooks/sandbox.sh
        build dune build -p ocaml_intrinsics -j 7".

#=== ERROR while compiling ocaml_intrinsics.v0.15~preview.124.35+330 ==========#
# context     2.0.8 | macos/x86_64 | ocaml-base-compiler.4.11.1 | https://ocaml.janestreet.com/opam-repository
# path        ~/Github/core_bench/_opam/.opam-switch/build/ocaml_intrinsics.v0.15~preview.124.35+330
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml_intrinsics -j 7
# exit-code   1
# env-file    ~/.opam/log/ocaml_intrinsics-49976-b1a82b.env
# output-file ~/.opam/log/ocaml_intrinsics-49976-b1a82b.out
### output ###
#           cc src/crc_stubs.o (exit 1)
# (cd _build/default/src && /usr/bin/cc -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT -O2 -fno-strict-aliasing -fwrapv -mpopcnt -mlzcnt -mbmi -mprfchw -mprefetchwt1 -g -I /Users/mando/Github/core_bench/_opam/lib/ocaml -o crc_stubs.o -c crc_stubs.c)
# crc_stubs.c:8:11: error: '__builtin_ia32_crc32di' needs target feature sse4.2
#    return __builtin_ia32_crc32di(initial, data);
#           ^
# 1 error generated.



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
┌─ The following actions failed
│ λ build ocaml_intrinsics v0.15~preview.124.35+330
└─ 
╶─ No changes have been performed
```

**Comment by gretay-js** [comment]
Thank you for reporting it. When the crc32 instruction is not available, it shouldn't stop the whole library from building! A fix would be to add software implementation of crc32.

**Comment by idkjs** [comment]
I have never heard of `crc32`, still it seems to be installed:

```sh
which crc32
/usr/bin/crc32
```

**Comment by denis631** [comment]
I am also having the same issue while building on MacOS with Intel i7
```
[ERROR] The compilation of ocaml_intrinsics failed at "/Users/denis.grebennicov/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml_intrinsics -j 11".

#=== ERROR while compiling ocaml_intrinsics.v0.15~preview.124.35+330 ==========#
# context     2.0.8 | macos/x86_64 | ocaml-base-compiler.4.11.1 | https://ocaml.janestreet.com/opam-repository
# path        ~/Documents/mongo_utils/_opam/.opam-switch/build/ocaml_intrinsics.v0.15~preview.124.35+330
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml_intrinsics -j 11
# exit-code   1
# env-file    ~/.opam/log/ocaml_intrinsics-43284-4a8073.env
# output-file ~/.opam/log/ocaml_intrinsics-43284-4a8073.out
### output ###
Done: 55/132 (jobs: 10)[...]
# (cd _build/default/src && /usr/local/opt/ccache/libexec/cc -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT -O2 -fno-strict-aliasing -fwrapv -mpopcnt -mlzcnt -mbmi -mprfchw -mprefetchwt1 -g -I /Users/denis.grebennicov/Documents/mongo_utils/_opam/lib/ocaml -o crc_stubs.o -c crc_stubs.c)
# crc_stubs.c:8:11: error: '__builtin_ia32_crc32di' needs target feature sse4.2
#    return __builtin_ia32_crc32di(initial, data);
#           ^
# 1 error generated.
```

**Comment by denis631** [comment]
Hi @idkjs, did you manage to solve the issue?

**Comment by xclerc** [comment]
It looks like the following patch does solve the issue (joint work with @gretay-js):

```
diff --git a/src/discover/discover.ml b/src/discover/discover.ml
index 7d687b8..d062512 100644
--- a/src/discover/discover.ml
+++ b/src/discover/discover.ml
@@ -81,6 +81,7 @@ let () =
            ; "-mlzcnt", prog_lzcnt
            ; "-mbmi", prog_tzcnt
            ; "-mcrc32", prog_crc32
+           ; "-msse4.2", prog_crc32
            ; "-mcrc32", prog_crc32_on_32bit_target
            ; "-mprfchw", prog_prefetchw
            ; "-mprefetchwt1", prog_prefetchwt1
```

**Comment by xclerc** [comment]
Actually, I misread the first message: the patch above is expected
to fix the issue on macOS/Intel, but not on macOS/M1.

**Comment by gretay-js** [comment]
@idkjs  I have a tentative fix for mac/m1 [here](https://github.com/janestreet/ocaml_intrinsics/compare/master...gretay-js:macos_crc32) but I can't get my hands on such a device to test it. Is there any chance you can give it a go?
By the way, the error message you posted on the issue description does looks like it's coming from macos/x86_64 not m1.

**Comment by gretay-js** [comment]
> > I have a tentative fix for mac/m1 [here](https://github.com/janestreet/ocaml_intrinsics/compare/master...gretay-js:macos_crc32) but I can't get my hands on such a device to test it. Is there any chance you can give it a go?
> 
> @gretay-js With the patch proposed in your fork I am able to install `ocaml_intrinsics` on an m1 mac.

Thank you for testing and confirming it! The fix is in the process of being released and will be included in the next preview.

**Comment by aalekseyev** [comment]
The fix should be in 7d7668a753c1ada0d1d79e7e1fa6a3eddb61d349.

**Comment by idkjs** [comment]
> @idkjs I have a tentative fix for mac/m1 [here](https://github.com/janestreet/ocaml_intrinsics/compare/master...gretay-js:macos_crc32) but I can't get my hands on such a device to test it. Is there any chance you can give it a go? By the way, the error message you posted on the issue description does looks like it's coming from macos/x86_64 not m1.

Sorry for the late reply. Was away. I will try the update and revert if it doesnt work.
